### PR TITLE
Add asset root on block generation - Closes #6750

### LINKED ISSUE #6750: Add assetsRoot when generating a block
### Description

Creation and validation of the new property `assetsRoot` must be added as described in [LIP](https://github.com/LiskHQ/lips/blob/55ac3d48fd06ae54ee41b5ad9d7c90e7b01fd510/proposals/lip-update_block_schema_and_block_processing.md#block-header).

- assetsRoot must be validated at `validate` function of `Block` class (lisk-chain) as described in LIP.
- Implement `getRoot` method in `BlockAssets` class (lisk-chain)
- Add `assetsRoot` before signing in `_generateBlock` in generator class (lisk-framework)

### Acceptance Criteria
- Add unit tests to ensure asset root is correct
- Update generator unit test

### Additional Information
Related Issue #6749

### PR DESCRIPTION
**Pull request by ishantiw**
### What was the problem?

This PR resolves #6750

### How was it solved?

🌱 Add block asset validation https://github.com/LiskHQ/lisk-sdk/commit/004e232749f795f7bb97a4ee200f6badb9c76914

- Expose registered module ids from state machine
- Check the size of data of an asset not exceeding the limit
- Check a module asset is not duplicated
- Check a module assets are sorted by id and are registered in state machine
  
🌱 Calculate asset root and add it to block header on block generation https://github.com/LiskHQ/lisk-sdk/commit/f9337ec686a9f12a00cdfae5b3d72fe4b092269c
  
♻️ Add _validateBlockAsset https://github.com/LiskHQ/lisk-sdk/commit/d810783965a107b45ad4f472d1091932bf85a743
  
✅ Add Block and BlockAssets validation tests https://github.com/LiskHQ/lisk-sdk/commit/89778d62fecd4df99279d9d810b811f7c9440550
  
✅ Add tests for BlockAssets validation in consensus and generator https://github.com/LiskHQ/lisk-sdk/commit/b9dea4f7a37f1f31accde2a2b5effa9a712e5dc4
  

### How was it tested?

Under `framework`,
`npm run test:unit consensus.spec`
`npm run test:unit generator.spec`

Under `lisk-chain`,
`npm run test block_asset.spec`
`npm run test block.spec`

### CODE REVIEW

File: elements/lisk-chain/src/block_assets.ts [hunk 1]
```diff
@@ -46,6 +54,10 @@ export class BlockAssets {
 		return this._assets.find(a => a.moduleID === moduleID)?.data;
 	}
 
+	public getAllAssets(): BlockAsset[] {
```
Comment (shuse2) [inline elements/lisk-chain/src/block_assets.ts:57]:
```suggestion
	public getAll(): BlockAsset[] {
```
I  think it's better not to repeat `assets.getAll()`

File: elements/lisk-chain/src/block.ts [hunk 2]
```diff
@@ -76,5 +77,26 @@ export class Block {
 		for (const tx of this.payload) {
 			tx.validate();
 		}
+
+		const assets = this.assets.getAllAssets();
+		let last = assets[0];
+		let i = 0;
+		for (const asset of assets) {
+			// Data size of each module should not be greater than max asset data size
+			if (asset.data.byteLength > MAX_ASSET_DATA_SIZE_BYTES) {
```
Comment (shuse2) [inline elements/lisk-chain/src/block.ts:86]: it's good point to  use `byteLength`.  We might need to  check where we use `length` for Buffer and  make it consistent

File: elements/lisk-chain/src/constants.ts [hunk 3]
```diff
@@ -28,3 +28,6 @@ export const GENESIS_BLOCK_TRANSACTION_ROOT = EMPTY_HASH;
 
 export const TAG_BLOCK_HEADER = createMessageTag('BH');
 export const TAG_TRANSACTION = createMessageTag('TX');
+
+// TODO: Actual size TBD
+export const MAX_ASSET_DATA_SIZE_BYTES = 64;
```
Comment (shuse2) [inline elements/lisk-chain/src/constants.ts:33]: I think genesis block will exceed this value.

File: elements/lisk-chain/src/block.ts [hunk 4]
```diff
@@ -76,5 +77,26 @@ export class Block {
 		for (const tx of this.payload) {
 			tx.validate();
 		}
+
```
Comment (shuse2) [inline elements/lisk-chain/src/block.ts:80]: I think assetroot and tx root needed to be validated  here.
However, to do this in sync func, we need to create function to calculate merkle root. Ill create  separate  issue for this

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: elements/lisk-chain/src/block.ts (hunk 2 above)
Comment (ishantiw) [reply elements/lisk-chain/src/block.ts:86]: we can create an issue for this or consider this in upcoming PRs

File: elements/lisk-chain/src/constants.ts (hunk 3 above)
Comment (ishantiw) [reply elements/lisk-chain/src/constants.ts:33]: what could be the safe value until we get the announced value?

File: elements/lisk-chain/src/block.ts (hunk 4 above)
Comment (ishantiw) [reply elements/lisk-chain/src/block.ts:80]: for trx root, if we do it here then we can remove its validation from the generator I think

Comment (ishantiw) [reply elements/lisk-chain/src/block.ts:80]: Will be resolved in the issue you created https://github.com/LiskHQ/lisk-sdk/issues/6886

File: elements/lisk-chain/test/unit/block.spec.ts [hunk 5]
```diff
@@ -53,5 +71,127 @@ describe('block', () => {
...
+					},
+				];
+				block = await createValidDefaultBlock({ assets: new BlockAssets(assets) });
+				// Act & assert
+				expect(() => block.validate()).toThrow(
+					`Module with ID ${assets[1].moduleID} has data size more than ${MAX_ASSET_DATA_SIZE_BYTES} bytes.`,
+				);
+			});
+
+			it(`should pass when asset data lenght is equal or less than ${MAX_ASSET_DATA_SIZE_BYTES}`, async () => {
```
Comment (mehmetegemen) [inline elements/lisk-chain/test/unit/block.spec.ts:95]: typo at `lenght`

**Review by mehmetegemen — COMMENTED** [review]
Left a very minor comment, great job 🎉

File: elements/lisk-chain/test/unit/block.spec.ts (hunk 5 above)
Comment (ishantiw) [reply elements/lisk-chain/test/unit/block.spec.ts:95]: seems like there were other places too where i did that typo 😅

**Review by shuse2 — APPROVED** [review]

**Review by mehmetegemen — APPROVED** [review]

**Pull request merged**